### PR TITLE
Rewrite to Typescript

### PR DESCRIPTION
**Pull request by JerryGreen**
`next-absolute-url` doesn't have Typescript definitions:

Screenshot 2020-01-13 at 18 47 20

What this PR does:

1. Rewrite to Typescript.
It has `index.ts` now, which is actual source code now, and by running `yarn build` it will generate `index.js` (which stays to be pretty much the same as earlier) and `index.d.ts` with type definitions

2. Remove `req.headers['x-forwarded-host']` as host source.
I don't receive this header myself, and also I haven't found it at node type definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/http.d.ts

3. Use `/^localhost(:\d+)?$/.test(host)` regular expression instead of `host.indexOf('localhost') > -1`.
No very much sense here but urls like `fancy-localhost.com` won't be treated as actual localhost now

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { IncomingMessage } from 'http'
```
Comment (jekrb) [inline index.ts]: This is more of my lack of familiarity with typescript. Does importing from `http` here have an effect on bundle size? Or does it get compiled away with typescript?

File: index.js [hunk 2]
```diff
@@ -1,16 +1,19 @@
-function absoluteUrl(req, setLocalhost) {
-  var protocol = 'https:';
-  var host = req ? (req.headers['x-forwarded-host'] || req.headers['host']) : window.location.host;
-  if (host.indexOf('localhost') > -1) {
-    if (setLocalhost) host = setLocalhost;
-    protocol = 'http:';
+'use strict'
+exports.__esModule = true
+function absoluteUrl(req, localhostAdress) {
```
Comment (jekrb) [inline index.js]: Small typo fix

```suggestion
function absoluteUrl(req, localhostAddress) {
```

File: index.js [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+  if (localhostAdress === void 0) {
```
Comment (jekrb) [inline index.js]: TIL about `void 0`, this is just another way to say `localhostAddress` is undefined right? Is it more typescript idiomatic to compare against void 0?

```suggestion
  if (localhostAddress === void 0) {
```

File: index.js [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
+    localhostAdress = 'localhost:3000'
```
Comment (jekrb) [inline index.js]:
```suggestion
    localhostAddress = 'localhost:3000'
```

File: index.js [hunk 5]
```diff
@@ -1,16 +1,19 @@
...
+  }
+  var protocol = 'https:'
+  var host = req ? req.headers['host'] : window.location.host
+  if (!host || /^localhost(:\d+)?$/.test(host)) {
+    host = localhostAdress
```
Comment (jekrb) [inline index.js]:
```suggestion
    host = localhostAddress
```

File: index.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+function absoluteUrl(
+  req?: IncomingMessage,
+  localhostAdress = 'localhost:3000'
```
Comment (jekrb) [inline index.ts]:
```suggestion
  localhostAddress = 'localhost:3000'
```

File: index.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+) {
+  let protocol = 'https:'
+  let host = req ? req.headers['host'] : window.location.host
+  if (!host || /^localhost(:\d+)?$/.test(host)) {
+    host = localhostAdress
```
Comment (jekrb) [inline index.ts]:
```suggestion
    host = localhostAddress
```

File: index.ts (hunk 1 above)
Comment (JerryGreen) [reply index.ts]: It does get it away from the bundle, but if you'll look into `index.d.ts`, you'll notice this row:

```ts
/// <reference types="node" />
```

So it assumes you have `@types/node` when using this package, which is normally true. At least it's definitely true for a NextJS project.

File: index.js (hunk 2 above)
Comment (JerryGreen) [reply index.js]: It's not a typo, actually. "Address" is a proper one

File: index.js (hunk 3 above)
Comment (JerryGreen) [reply index.js]: Lol, I haven't got attention to this earlier. Yea, look like it's analogue of `typeof localhostAdress === 'undefined'`. Just learned it today too lol. Turns out it's safer in some cases, here's explanation of it:

https://stackoverflow.com/a/19369074/3720305

**Comment by JerryGreen** [comment]
> 1. Would you recommend a version bump for a typescript rewrite?

Yea. I'll make a version bump commit right away

> 2. Would you be willing to add a test for these changes?

Yea, I'll think about this too now

**Comment by JerryGreen** [comment]
@jekrb I have a question actually, - why you've initially added `host.indexOf('localhost') > -1`, does it have any sense, really?

**Comment by JerryGreen** [comment]
@jekrb I think that's all. Have a look?

**Comment by jekrb** [comment]
@JerryGreen this looks really good to me. Thank you so much for writing the tests! Merging now. 🚀

**Comment by jekrb** [comment]
@JerryGreen this has been publish on npm `next-absolute-url@1.2.0` 🎉

File: index.js (hunk 2 above)
Comment (JerryGreen) [reply index.js]: @jekrb lol just noticed I indeed did type "Adress". I was so sure I didn't misspell this that I thought my spelling is "Address" while you're trying to say that "Adress" is proper one lol. That night was a weird one

Comment (jekrb) [reply index.js]: No worries :)

**Pull request merged**